internal/dataplane/kongstate/kongstate.go

```diff
@@ -330,7 +335,32 @@ func (ks *KongState) FillVaults(
 	}
 }
 
-func (ks *KongState) getPluginRelations() map[string]util.ForeignRelations {
+// TODO TRR copypastaed this out of internal/dataplane/translator/translate_utils.go, it needs to live in some package
+
+// getPermittedForReferenceGrantFrom takes a ReferenceGrant From (a namespace, group, and kind) and returns a map
+// from a namespace to a slice of ReferenceGrant Tos. When a To is included in the slice, the key namespace has a
+// ReferenceGrant with those Tos and the input From.
+func getPermittedForReferenceGrantFrom(
+	from gatewayapi.ReferenceGrantFrom,
+	grants []*gatewayapi.ReferenceGrant,
+) map[gatewayapi.Namespace][]gatewayapi.ReferenceGrantTo {
+	allowed := make(map[gatewayapi.Namespace][]gatewayapi.ReferenceGrantTo)
+	// loop over all From values in all grants. if we find a match, add all Tos to the list of Tos allowed for the
+	// grant namespace. this technically could add duplicate copies of the Tos if there are duplicate Froms (it makes
+	// no sense to add them, but it's allowed), but duplicate Tos are harmless (we only care about having at least one
+	// matching To when checking if a ReferenceGrant allows a reference)
+	for _, grant := range grants {
+		for _, otherFrom := range grant.Spec.From {
+			if reflect.DeepEqual(from, otherFrom) {
+				allowed[gatewayapi.Namespace(grant.ObjectMeta.Namespace)] = append(allowed[gatewayapi.Namespace(grant.ObjectMeta.Namespace)], grant.Spec.To...)
+			}
+		}
+	}
+
+	return allowed
+}
+
+func (ks *KongState) getPluginRelations(cacheStore store.Storer, log logr.Logger) map[string]util.ForeignRelations {
```

Comment:
This `func` got a bit lengthy with this change, didn't it? I would suggest we consider refactoring this (a tiny bit) so that at the very least the added code that iterate over `ks.Services` and `ks.Consumers` is extracted out. WDYT?

Reply:
This is the

> Ideally we'd do away with it

from OP. I played around with:

- Extracted the new logic to a separate function that accepts a `type addPluginRelationFunc func(namespace, plugin, identifier string)` arg. Having a standalone helper that accepts a dynamically-instantiated function bound to some dynamic scope `map[string]util.ForeignRelations` instance felt icky.
- Making `map[string]util.ForeignRelations` a defined type with methods to add relations. This also resulted in a mess.

So IDK exactly how I'd want to refactor it, but I don't think there's a simple way to. Was there a smaller change you were thinking of? We're already iterating over everything that can bind to plugins, but the additions mean those iterated tasks are much more complicated, and unfortunately not in a way that actually separates out cleanly AFAICT.

Reply:
Something like this could work (although it could still be improved):

https://github.com/Kong/kubernetes-ingress-controller/compare/feat/plugin-across...refactor-getPluginRelations?expand=1

https://github.com/Kong/kubernetes-ingress-controller/commit/c53cd3c5bfd1eb2e6241a21c6697351c3e5e713c